[Custom fields] Create custom fields plugin and first Strapi custom field: Color Picker

markkaylor opened the pull request:
### What does it do?

- Adds a new package to the monorepo `@strapi/custom-fields` 
- Creates a new custom field: Color Picker

### Why is it needed?

Color picker is the most requested custom field

### How to test it?

- Run `yarn setup`, then run the `examples/getstarted` app with `yarn develop --watch-admin`
- Select any content-type in the CTB, add a new field, select custom, select Color Picker (not Color!), save the content-type
- Head to your content-type in the content manager, create a new entity
- You should see an input for the custom field that matches the design system
- It should have a preview circle of the color and a hex value, it should default to white #FFFFFF
- It should show a ui when clicking the input
- It should update the hex value in the UI and on the toggle button preview
- You should be able to update the HEX on the UI's input and the change should show in the color saturation / hue 
- It should close when clicking outside the UI or when clicking the toggle button
- The entity should save to the db

gu-stav commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerIcon/index.js:
```diff
@@ -0,0 +1,6 @@
+import React from 'react';
+import icon from './icon.svg';
+
+const ColorPickerIcon = () => <img src={icon} />;
```
Could we give the the icon either: 1) an alt text 2) an aria-hidden attribute, if it is decorative only?

Also: should we try and move this icon into the DS? //cc @maevalienard

markkaylor replied:
I think actually here we should probably be rendering the svg inline to match the html structure of the other strapi attribute icons.

For the aria-hidden it should probably be on all attribute icons not just custom field icons, right? So maybe I should make that change in the [`AttributeIcon`](https://github.com/strapi/strapi/blob/features/custom-fields/packages/core/content-type-builder/admin/src/components/AttributeIcon/index.js) component, wdyt?

Moving the icon to the DS is fine by me, but I think custom field developers will probably reference this custom-field when building their own so it could be nice to show them that they don't have to use an icon from the DS and how to use an icon that is not in the DS. I guess that example could also just be in the documentation.

maevalienard replied:
The objective for now is not to allow external developers to use icons outside of our DS, that's why we added a bunch of random icons in the DS. So I'd rather not go that way, and keep it consistent with the strategy :)

markkaylor replied:
So should I add this icon to the DS or use a random one that already exists? @maevalienard could you recommend an icon to use?

gu-stav replied:
@markkaylor In case you add a new one, please sync with @HichamELBSI because he will do a release today :)

maevalienard replied:
It's already part of the DS under the name "Paint" then :)

markkaylor replied:
Awesome! thanks!

gu-stav commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,178 @@
+  height: 20px;
+  margin-right: 10px;
+  background-color: ${(props) => props.color};
+  border: 1px solid rgba(0, 0, 0, 0.1);
```
Are we sure we want another shade of grey?

markkaylor replied:
@maevalienard is there a value from [the DS colors](https://github.com/strapi/design-system/blob/main/packages/strapi-design-system/src/themes/colors.js) I could use?

gu-stav commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,178 @@
+    >
+      <Stack spacing={1}>
+        <Flex>
+          <FieldLabel required={required}>{formatMessage(intlLabel)}</FieldLabel>
```
I think you could pass `labelAction` as `action` prop to `FieldLabel` instead. This way you make sure it uses the proper `Stack` spacing.

gu-stav commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,178 @@
+              {color}
+            </Typography>
+          </Flex>
+          <CarretDown />
```
I think this icon needs to be visually hidden (e.g. through `aria-hidden`).

gu-stav commented on packages/plugins/custom-fields/admin/src/hooks/useOnClickOutside.js:
```diff
@@ -0,0 +1,15 @@
+
+const useOnClickOutside = (ref, listener) => {
+  useEffect(() => {
+    document.addEventListener('mousedown', listener);
+    document.addEventListener('touchstart', listener);
```
Why don't you use `click` instead? I think it would trigger in both cases.

gu-stav commented on packages/plugins/custom-fields/admin/src/translations/en.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "input.format": "HEX",
```
Why doesn't this have the `color-picker` NS?

maevalienard commented:
Hey Mark! Just checking the PR, and I was wondering about the level of granularity we have regarding the design so it could match our DS?

HichamELBSI commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,178 @@
+  value: PropTypes.string,
+};
+
+export default ColorPickerInput;
```
Could you write some test for this component please 😅

markkaylor commented:
@Feranchz I added regex and required field to the custom field. The config isn't set by default so you have to add the regex when adding your custom field to the content type in the advanced tab `^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$`. I don't think we should handle this directly in the component itself but maybe we could consider default values for these inputs 🤷‍♂️

markkaylor commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerIcon/index.js:
```diff
@@ -1,6 +1,26 @@
+  border-radius: 3px;
+  display: flex;
+  justify-content: center;
+  align-items: center;
```
Yes good point, my bad, updated.

maevalienard commented:
Hello 👋 Here are my two cents and some questions as well:
- Would it be possible to have a rendering similar to the design for the color picked?
![Capture d’écran 2022-08-31 à 15 56 14](https://user-images.githubusercontent.com/54801383/187696816-2072cd59-07a3-4047-ae40-51a3d48147b7.png)

- I don't think the description "Using color format rgba" or "Using color format hex" is needed. 

- When we click on the color, I think that'd be great to display the right selector and disable (don't show it at all would be preferable though) the other one(s). Otherwise, the format requested in the CTB doesn't make real sense. In case, the disable state in not feasible, I think we should remove the format request input in the CTB. 

- There's another comment about that, but for the dropdown, the border color should be neutral200 with a border-radius of 4px and a popUp shadow behind. 

- I think the inputs should be the ones that we have in the Design System so they also would have the right font and color inside (neutral800 with an Omega font).

remidej commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,196 @@
+      hint={description && formatMessage(description)}
+    >
+      <Stack spacing={1}>
+        <Flex>
```
You'll want something like

```js
{labelAction && <Box paddingLeft={1}>{labelAction}</Box>}
```

to that Flex, so that i18n can inject the localized icons

markkaylor replied:
I believe it is handled by passing `labelAction` to the `action` prop on `FieldLabel`, [Suggestion from Gustav](https://github.com/strapi/strapi/pull/14212#discussion_r954913069)

markkaylor replied:
So I think can remove this Flex also now that I think about

remidej replied:
awesome

remidej commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,198 @@
+        </Flex>
+        <ColorPickerToggle
+          ref={colorPickerButtonRef}
+          aria-label="Color picker toggle"
```
We may want to translate this

markkaylor commented:
Thanks for the feedback @maevalienard!

> I'm not able to check in the code, but is the background of this icon using a color from the DS? It seems that in Dark mode, the background color is not the same than for the other CTB icons. Just in case, the background should have a 
primary100 color + a primary200 border color.

I have these values set for backgorund and borderColor and they seem to be coming through.

Background for example:
- For dark mode primary100: `#181826`
- For light mode primary100: `#f0f0ff`

I think the other CTB icons have their colors "hard-coded" so they don't change between light and dark mode. I guess I could do the same thing here by hard coding a consistent color value for both light and dark themes (knowing that eventually the idea would be to add a new icon to the DS and the update `Icon` component to handle the colors) 

> Do we have the possibility to display the dropdown the same width than the input? I don't know if it's responsive or not. If not, don't bother :)

We can display it full width but I think it looks a bit strange when using a big input. We would need to handle the height of the saturation box for different screen sizes and I feel like the input looks a bit lost over there in the corner. Up to you though!
![Screenshot 2022-09-01 at 10 53 06](https://user-images.githubusercontent.com/26598053/187876745-143244de-ce9d-4b89-bbf5-f730cd7d07eb.png)

> Very random question but is there a reason why we cannot go lower than that in the shade of the color palette?

Fixed!

> The "HEX" word in the dropdown should have a neutral600 color

Fixed!

remidej commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,198 @@
+            <FocusTrap onEscape={() => setShowColorPicker(false)}>
+              <ColorPicker
+                color={color}
+                onChange={(hexValue) =>
```
Did `onChange={onChange}` not work?

markkaylor replied:
It did not, if I remember correctly but I'll try again

remidej commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,198 @@
+                  aria-label="Color picker input"
+                  style={styleUppercase}
+                  value={value}
+                  placeholder="#FFFFFF"
```
I think it may be confusing to see white by default. Maybe we could think of a differnet "no color" state

remidej commented on packages/plugins/custom-fields/admin/src/components/ColorPicker/ColorPickerInput/index.js:
```diff
@@ -0,0 +1,198 @@
+`;
+
+const ColorPickerPopover = styled(Popover)`
+  padding: ${({ theme }) => theme.spaces[1]};
```
FYI on firefox the handles are slightly cropped out as they move out of the container
<img width="451" alt="Screenshot 2022-09-01 at 11 05 40" src="https://user-images.githubusercontent.com/8087692/187880721-b344251b-0317-4802-8c75-f73bd5b60cca.png">

markkaylor replied:
Yeah I think I have the same thing on chrome. The only thing I can think to do is increase the padding of the popover so it isn't cut off, could we do that @maevalienard? I don't really have much control over the pointer's  position unfortunately
![Screenshot 2022-09-01 at 12 01 45](https://user-images.githubusercontent.com/26598053/187888110-8352f2d2-0097-4f02-8251-f226d64942d5.png)

remidej replied:
Otherwise we may be able to hack something using CSS I think, like with negative negative margins, z-index or something

maevalienard replied:
Yes that would work, you could add 4 additional px so that would fit the grid of the DS :)

markkaylor replied:
ok great already updated :)

remidej commented on packages/plugins/custom-fields/package.json:
```diff
@@ -0,0 +1,37 @@
+  },
+  "dependencies": {},
+  "peerDependencies": {
+    "@strapi/strapi": "^4.0.0"
```
Will require `^4.4.0`

remidej reviewed (changes requested):
It looks great, congrats!

I personally think that we can remove the mycustomfields plugin now that we have this official custom-fields plugin. But that's another topic and could be done later

maevalienard commented:
> I think the other CTB icons have their colors "hard-coded" so they don't change between light and dark mode. I guess I could do the same thing here by hard coding a consistent color value for both light and dark themes (knowing that eventually the idea would be to add a new icon to the DS and the update Icon component to handle the colors)

That's good to know. I think this might be a problem in long term. Because on our side, we can find solutions, but we'll still have an issue with the custom fields from external developers as well... :/ I'll discuss with Emilie what she thinks. 

> We can display it full width but I think it looks a bit strange when using a big input. We would need to handle the height of the saturation box for different screen sizes and I feel like the input looks a bit lost over there in the corner. Up to you though!

Okay let's keep it without the full width size then! Thanks :D

remidej reviewed (approved):
Merging this as all the feedback should be addressed and Mark is OOO
